### PR TITLE
GatherGrad optimization

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
The existing implementation of the GatherGrad CUDA kernel does not do work in a very parallel manner for certain inputs which can lead to poor performance.

The computation essentially involves multiple summations. The values are gathered from the input and the sums are scattered to the output.

Previously, each sum was computed by a single thread. If there is an instance of a summation of a large number of values, it can significantly impact the overall kernel execution time.

The updated version has an alternate implementation which splits the sums into partial sums which get accumulated together later. This allows for more parallelism. A significant downside is that the alternate implementation requires CPU and GPU synchronization because intermediate GPU results are required by the CPU computation. The original implementation outperformed the alternate for certain inputs (e.g., where the maximum number of values in a sum was not large), so the updated version chooses between them based on the input. The input analysis has some overhead which shows up in the kernel benchmark measurements below.

The implementation was adapted from [PyTorch](https://github.com/pytorch/pytorch/blob/b186831c08e0e4e447eedb8a5cfab582995d37f9/aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu).

Kernel benchmark results:

gather input shape | indices shape | axis | baseline (microseconds) | updated (microseconds) | (baseline-updated)/baseline
-- | -- | -- | -- | -- | --
(32, 1024) | (128, 128) | 0 | 692.7676 | 437.976 | 0.367788
(32, 1024) | (128, 128) | 1 | 584.4383 | 617.7265 | -0.05696
(32, 1024) | (256, 128) | 0 | 1079.339 | 548.1611 | 0.492132
(32, 1024) | (256, 128) | 1 | 787.354 | 625.4328 | 0.205652
(32, 1024) | (512, 128) | 0 | 1850.218 | 769.2486 | 0.584239
(32, 1024) | (512, 128) | 1 | 1210.273 | 815.8086 | 0.32593
(32, 2048) | (128, 128) | 0 | 762.9307 | 543.3546 | 0.287806
(32, 2048) | (128, 128) | 1 | 522.4106 | 552.0353 | -0.05671
(32, 2048) | (256, 128) | 0 | 1226.961 | 756.6984 | 0.383274
(32, 2048) | (256, 128) | 1 | 666.1524 | 719.5856 | -0.08021
(32, 2048) | (512, 128) | 0 | 2149.223 | 1188.688 | 0.446922
(32, 2048) | (512, 128) | 1 | 983.3189 | 896.4058 | 0.088387
(32, 4096) | (128, 128) | 0 | 910.7237 | 746.6869 | 0.180117
(32, 4096) | (128, 128) | 1 | 501.7036 | 526.7084 | -0.04984
(32, 4096) | (256, 128) | 0 | 1528.694 | 1163.213 | 0.239081
(32, 4096) | (256, 128) | 1 | 634.188 | 676.4797 | -0.06669
(32, 4096) | (512, 128) | 0 | 2757.369 | 2000.065 | 0.274647
(32, 4096) | (512, 128) | 1 | 956.294 | 1018.83 | -0.06539

**Motivation and Context**
Performance.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/cu_inc/common.cuh [hunk 1]
```diff
@@ -230,7 +230,7 @@ __device__ __inline__ T _Gelu(T a) {
 #endif
 
 template <class INT, class INT2>
-static INT CeilDiv(INT a, INT2 b)  // ceil(a/b)
+inline __host__ __device__ INT CeilDiv(INT a, INT2 b)  // ceil(a/b)
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/cu_inc/common.cuh:233]: >__device__ [](start = 16, length = 10)

if you use this function in the device, I am wondering if it should be replaced with fast_divmod

**Review by SherlockNoMad — APPROVED** [review]

**Review by ytaous — DISMISSED** [review]

**Comment by weixingzhang** [comment]
Out of curiosity, based on the PR description, does it mean the perf gain with updated version is very small?

File: onnxruntime/core/providers/cuda/cu_inc/accumulation_type.cuh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+struct AccumulationType;
+template <>
+struct AccumulationType<half> { using type = float; };
+template <>
+struct AccumulationType<float> { using type = float; };
+template <>
+struct AccumulationType<double> { using type = double; };
+
+template <typename T>
+using AccumulationType_t = typename AccumulationType<T>::type;
```
Comment (weixingzhang) [inline onnxruntime/core/providers/cuda/cu_inc/accumulation_type.cuh:22]: struct AccumulateType has been defined in a few files (bias_softmax_impl.cu, softmax.h and reduction_all.cc). Could you help refactor the code to keep only one and remove others?

**Comment by edgchen1** [comment]
It depends on the inputs. The measurements in the table range from -8% to 58% improvement.

---
In reply to: [714241443](https://github.com/microsoft/onnxruntime/pull/5524#issuecomment-714241443) [](ancestors = 714241443)

**Review by weixingzhang — APPROVED** [review]
